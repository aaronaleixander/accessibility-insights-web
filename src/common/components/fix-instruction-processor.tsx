// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { RecommendColor } from 'common/components/recommend-color';
import * as React from 'react';
import * as styles from './cards/fix-instruction-color-box.scss';

type ColorMatch = {
    splitIndex: number;
    colorHexValue: string;
};

export class FixInstructionProcessor {
    private readonly colorValueMatcher = `(#[0-9a-f]{6})`;
    private readonly foregroundColorText = 'foreground color: ';
    // the following warnings can be disabled because the values are actually constant strings and the string template is used merely for ease of reading
    // eslint-disable-next-line security/detect-non-literal-regexp
    private readonly foregroundRegExp = new RegExp(
        `${this.foregroundColorText}${this.colorValueMatcher}`,
        'i',
    );
    private readonly backgroundColorText = 'background color: ';
    // eslint-disable-next-line security/detect-non-literal-regexp
    private readonly backgroundRegExp = new RegExp(
        `${this.backgroundColorText}${this.colorValueMatcher}`,
        'i',
    );

    private readonly foregroundRecommendedColorText = 'Use foreground color: ';
    private readonly foregroundRecommendedRegExp = new RegExp(
        `${this.foregroundRecommendedColorText}${this.colorValueMatcher}`,
        'i',
    );
    private readonly backgroundRecommendedColorText = 'Use background color: ';
    private readonly backgroundRecommendedRegExp = new RegExp(
        `${this.backgroundRecommendedColorText}${this.colorValueMatcher}`,
        'i',
    );

    private readonly contrastRatioText = 'Expected contrast ratio of ';
    private readonly contrastRatioRegExp = new RegExp(`${this.contrastRatioText}`, 'i');
    private readonly originalMiddleSentence = ' and the original foreground color: ';

    private readonly recommendEndSentence = ' to meet a contrast ratio of #.##:1';

    public process(fixInstruction: string, recommendation: RecommendColor): JSX.Element {
        fixInstruction += '.';
        const matches = this.getColorMatches(fixInstruction);

        if (matches.length === 2 && fixInstruction != null) {
            fixInstruction = this.getColorRecommendation(fixInstruction, matches, recommendation);
        }
        return this.splitFixInstruction(fixInstruction, matches);
    }

    private getColorRecommendation(
        fixInstruction: string,
        matches: ColorMatch[],
        recommendation: RecommendColor,
    ) {
        let contrastRatio: number = 4.5;
        const regularExpExpectation = this.contrastRatioRegExp.exec(fixInstruction);
        if (regularExpExpectation != null) {
            const indexContrast = regularExpExpectation.index;
            const contrastIndex = fixInstruction.substring(
                indexContrast + this.contrastRatioText.length,
            );
            contrastRatio = parseFloat(contrastIndex.substring(0, contrastIndex.indexOf(':')));
        }

        if (recommendation.sentence !== 'Color suggestion text') {
            fixInstruction += recommendation.getRecommendColor(
                matches[0].colorHexValue,
                matches[1].colorHexValue,
                contrastRatio,
            );
            this.getRecommendedColorMatches(fixInstruction, matches);
        }

        return fixInstruction;
    }

    private getColorMatches(fixInstruction: string): ColorMatch[] {
        const foregroundMatch = this.getColorMatch(fixInstruction, this.foregroundRegExp);
        const backgroundMatch = this.getColorMatch(fixInstruction, this.backgroundRegExp);

        return [foregroundMatch, backgroundMatch].filter(match => match != null) as ColorMatch[];
    }

    private getRecommendedColorMatches(fixInstruction: string, matches: ColorMatch[]) {
        const foregroundRecommededMatch = this.getColorMatch(
            fixInstruction,
            this.foregroundRecommendedRegExp,
        );
        if (foregroundRecommededMatch !== null) {
            matches.push(foregroundRecommededMatch);
            matches.push(this.getOriginalColorMatch(fixInstruction, foregroundRecommededMatch));
        }

        const backgroundRecommededMatch = this.getColorMatch(
            fixInstruction,
            this.backgroundRecommendedRegExp,
        );
        if (backgroundRecommededMatch !== null) {
            matches.push(backgroundRecommededMatch);
            matches.push(this.getOriginalColorMatch(fixInstruction, backgroundRecommededMatch));
        }
    }

    private getColorMatch(fixInstruction: string, colorRegex: RegExp): ColorMatch | null {
        if (!colorRegex.test(fixInstruction)) {
            return null;
        }

        const match = colorRegex.exec(fixInstruction);

        if (match == null || match[1] == null) {
            return null;
        }

        const colorHexValue = match[1];

        if (
            colorRegex === this.foregroundRecommendedRegExp ||
            colorRegex === this.backgroundRecommendedRegExp
        ) {
            return {
                splitIndex:
                    match.index + this.foregroundRecommendedColorText.length + colorHexValue.length,
                colorHexValue,
            };
        }

        return {
            splitIndex: match.index + this.foregroundColorText.length + colorHexValue.length,
            colorHexValue,
        };
    }

    private getOriginalColorMatch(fixInstruction: string, match: ColorMatch): ColorMatch {
        const split = match.splitIndex + this.originalMiddleSentence.length;
        return {
            splitIndex: split + match.colorHexValue.length,
            colorHexValue: fixInstruction.substring(split, split + match.colorHexValue.length),
        };
    }

    private splitFixInstruction(fixInstruction: string, matches: ColorMatch[]): JSX.Element {
        const sortedMatches = matches.sort((a, b) => a.splitIndex - b.splitIndex);
        if (sortedMatches.length === 0) {
            return <>{fixInstruction}</>;
        }

        const insertionIndex = 0;
        const keyIndex = 0;

        const result: JSX.Element[] = [];
        if (matches.length >= 2) {
            this.bulletAndColorBoxOutput(
                sortedMatches,
                fixInstruction,
                insertionIndex,
                result,
                keyIndex,
            );
        }

        return (
            <>
                <span aria-hidden="true">{result}</span>
                <span className={styles.screenReaderOnly}>{fixInstruction}</span>
            </>
        );
    }

    private bulletAndColorBoxOutput(
        sortedMatches: ColorMatch[],
        fixInstruction: string,
        insertionIndex: number,
        result: JSX.Element[],
        keyIndex: number,
    ) {
        for (let i: number = 0; i < sortedMatches.length; i++) {
            const match: ColorMatch = sortedMatches[i];
            if (i === 0) {
                insertionIndex = 0;
            } else {
                insertionIndex =
                    match.splitIndex -
                    this.foregroundRecommendedColorText.length -
                    match.colorHexValue.length -
                    1;
                if (i > 2) {
                    insertionIndex -= 3;
                }
            }
            let colorEndIndex = match.splitIndex - match.colorHexValue.length;
            const beforeColorBox = fixInstruction.substring(insertionIndex, colorEndIndex);
            insertionIndex = colorEndIndex;
            i++;
            colorEndIndex = sortedMatches[i].splitIndex - match.colorHexValue.length;
            const middleSubstring = fixInstruction.substring(insertionIndex, colorEndIndex);
            insertionIndex = colorEndIndex;
            let endSubstring = '';
            if (i === 1) {
                endSubstring = fixInstruction.substring(
                    insertionIndex,
                    fixInstruction.indexOf(this.contrastRatioText) +
                        this.contrastRatioText.length +
                        6,
                );
                result.push(
                    <span key={`instruction-split-${keyIndex++}`}>
                        {this.addBulletPoint(
                            keyIndex,
                            beforeColorBox,
                            match,
                            middleSubstring,
                            sortedMatches,
                            i,
                            endSubstring,
                        )}
                    </span>,
                );
            } else {
                colorEndIndex = sortedMatches[i].splitIndex + this.recommendEndSentence.length;
                endSubstring = fixInstruction.substring(insertionIndex, colorEndIndex);
                result.push(
                    <span key={`instruction-split-${keyIndex++}`}>
                        <ul key={`instruction-split-${keyIndex++}`}>
                            <li key={`instruction-split-${keyIndex++}`}>
                                {this.addBulletPoint(
                                    keyIndex,
                                    beforeColorBox,
                                    match,
                                    middleSubstring,
                                    sortedMatches,
                                    i,
                                    endSubstring,
                                )}
                            </li>
                        </ul>
                    </span>,
                );
            }
        }
    }

    private addBulletPoint(
        keyIndex: number,
        beforeColorBox: string,
        match: ColorMatch,
        middleSubstring: string,
        sortedMatches: ColorMatch[],
        i: number,
        endSubstring: string,
    ): JSX.Element {
        return (
            <span key={`instruction-split-${keyIndex++}`}>
                {beforeColorBox}
                {this.createColorBox(match.colorHexValue, keyIndex++)}
                {middleSubstring}
                {this.createColorBox(sortedMatches[i].colorHexValue, keyIndex++)}
                {endSubstring}
            </span>
        );
    }

    private createColorBox(colorHexValue: string, keyIndex: number): JSX.Element {
        return (
            <span
                key={`instruction-split-${keyIndex}`}
                className={styles.fixInstructionColorBox}
                style={{ backgroundColor: colorHexValue }}
            />
        );
    }
}
