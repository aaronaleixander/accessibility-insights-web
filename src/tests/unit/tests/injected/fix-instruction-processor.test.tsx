// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { FixInstructionProcessor } from 'common/components/fix-instruction-processor';
import { RecommendColor } from 'common/components/recommend-color';
import { Mock, Times } from 'typemoq';

describe('FixInstructionProcessor', () => {
    let testSubject: FixInstructionProcessor;
    let recommendationStub: RecommendColor;

    beforeEach(() => {
        testSubject = new FixInstructionProcessor();
        recommendationStub = {
            sentence: 'Color suggestion text',
        } as RecommendColor;
    });

    test('no background nor foreground on the message', () => {
        const fixInstruction =
            'there is nothing that will trigger the processor to change the fix instruction here';

        const result = testSubject.process(fixInstruction, recommendationStub);
        expect(result).toMatchSnapshot();
    });

    test('foreground color on the message', () => {
        const fixInstruction =
            'color contrast of 2.52 (foreground color: #8a94a8; nothing else to match here)';

        const result = testSubject.process(fixInstruction, recommendationStub);

        expect(result).toMatchSnapshot();
    });

    test('background color on the message', () => {
        const fixInstruction =
            'color contrast of 2.52 (background color: #8a94a8; nothing else to match here)';

        const result = testSubject.process(fixInstruction, recommendationStub);

        expect(result).toMatchSnapshot();
    });

    test('foreground and background color on the message (mind the order)', () => {
        const fixInstruction =
            'Element has insufficient color contrast of 2.52 (foreground color: #8a94a8, background color: #e7ecd8, font size: 12.0pt, font weight: normal). Expected contrast ratio of 4.5:1';

        const result = testSubject.process(fixInstruction, recommendationStub);

        expect(result).toMatchSnapshot();
    });

    test('background and foreground on the message (mind the order)', () => {
        const fixInstruction =
            'Element has insufficient color contrast of 2.52 (background color: #8a94a8, foreground color: #e7ecd8, font size: 12.0pt, font weight: normal). Expected contrast ratio of 4.5:1';

        const result = testSubject.process(fixInstruction, recommendationStub);

        expect(result).toMatchSnapshot();
    });

    test('we assume only one instance of fore/background color, second instance will be ignored', () => {
        const fixInstruction =
            'first foreground color: #ffffff, second foreground: #000000, first background color: #AAAAAA, second background color: #bbBBbb';

        const result = testSubject.process(fixInstruction, recommendationStub);

        expect(result).toMatchSnapshot();
    });

    test('when we have nothing to process', () => {
        const fixInstruction = 'nothing to process';

        const result = testSubject.process(fixInstruction, recommendationStub);

        expect(result).toMatchSnapshot();
    });

    test('recommendColor non-stub', () => {
        const fixInstruction =
            'Element has insufficient color contrast of 4.48 (foreground color: #fefefe, background color: #21809d, font size: 10.8pt (14.4px), font weight: normal). Expected contrast ratio of 4.5:1';
        
        const getRecommendColorReturnValue = 
            'Use foreground color: #ffffff and the original background color: #21809d to meet a contrast ratio of 4.53:1. Or use background color: #1f7995 and the original foreground color: #fcfcfc to meet a contrast ratio of 4.84:1.';
        const recommendColorMock = Mock.ofType(RecommendColor);

        recommendColorMock
            .setup(rc => rc.getRecommendColor('#fefefe', '#21809d', 4.5))
            .returns(() => getRecommendColorReturnValue)
            .verifiable(Times.once());
        

        const result = testSubject.process(fixInstruction, recommendColorMock.object);

        expect(result).toMatchSnapshot();
        recommendColorMock.verifyAll();
    });
});
