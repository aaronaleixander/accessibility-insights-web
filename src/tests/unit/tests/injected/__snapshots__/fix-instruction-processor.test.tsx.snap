// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixInstructionProcessor background and foreground on the message (mind the order) 1`] = `
<React.Fragment>
  <span
    aria-hidden="true"
  >
    <span>
      <span>
        Element has insufficient color contrast of 2.52 (background color: 
        <span
          className="fixInstructionColorBox"
          style={
            Object {
              "backgroundColor": "#8a94a8",
            }
          }
        />
        #8a94a8, foreground color: 
        <span
          className="fixInstructionColorBox"
          style={
            Object {
              "backgroundColor": "#e7ecd8",
            }
          }
        />
        #e7ecd8, font size: 12.0pt, font weight: normal). Expected contrast ratio of 4.5:1.
      </span>
    </span>
  </span>
  <span
    className="screenReaderOnly"
  >
    Element has insufficient color contrast of 2.52 (background color: #8a94a8, foreground color: #e7ecd8, font size: 12.0pt, font weight: normal). Expected contrast ratio of 4.5:1.
  </span>
</React.Fragment>
`;

exports[`FixInstructionProcessor background color on the message 1`] = `
<React.Fragment>
  <span
    aria-hidden="true"
  />
  <span
    className="screenReaderOnly"
  >
    color contrast of 2.52 (background color: #8a94a8; nothing else to match here).
  </span>
</React.Fragment>
`;

exports[`FixInstructionProcessor foreground and background color on the message (mind the order) 1`] = `
<React.Fragment>
  <span
    aria-hidden="true"
  >
    <span>
      <span>
        Element has insufficient color contrast of 2.52 (foreground color: 
        <span
          className="fixInstructionColorBox"
          style={
            Object {
              "backgroundColor": "#8a94a8",
            }
          }
        />
        #8a94a8, background color: 
        <span
          className="fixInstructionColorBox"
          style={
            Object {
              "backgroundColor": "#e7ecd8",
            }
          }
        />
        #e7ecd8, font size: 12.0pt, font weight: normal). Expected contrast ratio of 4.5:1.
      </span>
    </span>
  </span>
  <span
    className="screenReaderOnly"
  >
    Element has insufficient color contrast of 2.52 (foreground color: #8a94a8, background color: #e7ecd8, font size: 12.0pt, font weight: normal). Expected contrast ratio of 4.5:1.
  </span>
</React.Fragment>
`;

exports[`FixInstructionProcessor foreground color on the message 1`] = `
<React.Fragment>
  <span
    aria-hidden="true"
  />
  <span
    className="screenReaderOnly"
  >
    color contrast of 2.52 (foreground color: #8a94a8; nothing else to match here).
  </span>
</React.Fragment>
`;

exports[`FixInstructionProcessor no background nor foreground on the message 1`] = `
<React.Fragment>
  there is nothing that will trigger the processor to change the fix instruction here.
</React.Fragment>
`;

exports[`FixInstructionProcessor recommendColor non-stub 1`] = `
<React.Fragment>
  <span
    aria-hidden="true"
  >
    <span>
      <span>
        Element has insufficient color contrast of 4.48 (foreground color: 
        <span
          className="fixInstructionColorBox"
          style={
            Object {
              "backgroundColor": "#fefefe",
            }
          }
        />
        #fefefe, background color: 
        <span
          className="fixInstructionColorBox"
          style={
            Object {
              "backgroundColor": "#21809d",
            }
          }
        />
        #21809d, font size: 10.8pt (14.4px), font weight: normal). Expected contrast ratio of 4.5:1.
      </span>
    </span>
    <span>
      <ul>
        <li>
          <span>
            .Use foreground color: 
            <span
              className="fixInstructionColorBox"
              style={
                Object {
                  "backgroundColor": "#ffffff",
                }
              }
            />
            #ffffff and the original background color: 
            <span
              className="fixInstructionColorBox"
              style={
                Object {
                  "backgroundColor": "#21809d",
                }
              }
            />
            #21809d to meet a contrast ratio of 4.53:1
          </span>
        </li>
      </ul>
    </span>
    <span>
      <ul>
        <li>
          <span>
             Or use background color: 
            <span
              className="fixInstructionColorBox"
              style={
                Object {
                  "backgroundColor": "#1f7995",
                }
              }
            />
            #1f7995 and the original foreground color: 
            <span
              className="fixInstructionColorBox"
              style={
                Object {
                  "backgroundColor": "#fcfcfc",
                }
              }
            />
            #fcfcfc to meet a contrast ratio of 4.84:1
          </span>
        </li>
      </ul>
    </span>
  </span>
  <span
    className="screenReaderOnly"
  >
    Element has insufficient color contrast of 4.48 (foreground color: #fefefe, background color: #21809d, font size: 10.8pt (14.4px), font weight: normal). Expected contrast ratio of 4.5:1.Use foreground color: #ffffff and the original background color: #21809d to meet a contrast ratio of 4.53:1. Or use background color: #1f7995 and the original foreground color: #fcfcfc to meet a contrast ratio of 4.84:1.
  </span>
</React.Fragment>
`;

exports[`FixInstructionProcessor we assume only one instance of fore/background color, second instance will be ignored 1`] = `
<React.Fragment>
  <span
    aria-hidden="true"
  >
    <span>
      <span>
        first foreground color: 
        <span
          className="fixInstructionColorBox"
          style={
            Object {
              "backgroundColor": "#ffffff",
            }
          }
        />
        #ffffff, second foreground: #000000, first background color: 
        <span
          className="fixInstructionColorBox"
          style={
            Object {
              "backgroundColor": "#AAAAAA",
            }
          }
        />
         second foreground: #000000, first background color: 
      </span>
    </span>
  </span>
  <span
    className="screenReaderOnly"
  >
    first foreground color: #ffffff, second foreground: #000000, first background color: #AAAAAA, second background color: #bbBBbb.
  </span>
</React.Fragment>
`;

exports[`FixInstructionProcessor when we have nothing to process 1`] = `
<React.Fragment>
  nothing to process.
</React.Fragment>
`;
